### PR TITLE
Rename galois to katana

### PR DESCRIPTION
**Pull request by ddn0**
# General philosophy
- Galois vs Katana. "Katana" refers to the general project. "Galois" refers the shared-memory parallel programming library, e.g., parallel loop iterators, data structures (but not analytics algorithms).
- One namespace (more or less). Having too many nested namespaces isn't recommended for C++ and many of the previous namespace distinctions like `galois::runtime` `galois::worklists` `galois::graphs` don't make sense to users of libgalois anyway.
- Provide space for subcomponents. We have several components like distributed execution, querying, tools, analytics routines that would benefit from consistent building, packaging and distribution.

# Solution
- Rename `galois::` namespace to `katana::` update include paths to match
- Flatten most nested `galois::foo` namespaces to just `katana::`. The notable exceptions are `galois::gstl` and `internal`. (Given the relatively infrequent use of `galois::gstl` and upcoming STL support for parallel algorithms, it might be reasonable to remove `galois::gstl` altogether in the future.)
- Update go, cmake, python, python build files to match

# Why not `s/galois/katana/g` (preserving case)?

Previously "Galois" referred to (1) the project, (2) the shared-memory programming library and (3) a general philosophy about thinking about parallel programming, (4) the brand. Each use of "Galois" needs to be examined for its specific meaning.

# Things that are explicitly out of scope
- Human text. This will require more sophisticated understanding of the meaning of "galois" and "katana" than what can be accomplished with regular expressions. It also is limited to a subset of files. #17 tracks this follow on work. 
- Style guide. Suggestions to future developers on the general structure, when/where/how to add subcomponents, etc. (see #17)
- License text

# Pending work
- [x] flatten `galois::substrate` to `katana`
- [x] rename CMake target `galois_shmem` target to `katana_galois`
- [x] move querying files from `include/querying` to `include/katana/querying` (and update include paths to match)

### CODE REVIEW

**Comment by arthurp** [comment]
Real review tomorrow night I hope, but a thought for now.

What is the `g` in `gInfo` (and other `gIO` functions)? Is it "galois"? Should we change it to something else? It cannot be `k` (because that's the constant prefix), but we could still rename the functions somehow. (I would love to kill those functions, but that's way out of scope for this PR.)

**Comment by ddn0** [comment]
> What is the g in gInfo (and other gIO functions)? Is it "galois"? Should we change it to something else? It cannot be k (because that's the constant prefix), but we could still rename the functions somehow. (I would love to kill those functions, but that's way out of scope for this PR.)

The "g" does stand for galois. New code uses Logging.h which basically supersedes `gInfo` and co., and a long-term plan is to remove gIO at some point.

Some context for this PR. I mainly focused on mechanical changes. Anything more nuanced probably should be happen after this PR lands.

File: Python.md
Comment (arthurp) [inline Python.md]: Replace `galois-env` with `katana-env`.

File: docs/introduction.dox
Comment (arthurp) [inline docs/introduction.dox:59]: You seem to have left "Galois" in the text of the documentation repeatedly (here and in other documentation files). Is this intentional? What's your rational? If you want to leave it I think we should add a blurb to the documentation expaining that we use the term "Galois" for the model and "Katana" for the library or whatever we actually do.

File: libgalois/include/katana/BufferedGraph.h
Comment (arthurp) [inline libgalois/include/katana/BufferedGraph.h:31]: Where did this come from?

File: .github/workflows/python.yaml
Comment (arthurp) [inline .github/workflows/python.yaml]: There are a number of "galois" still in this file.

**Review by arthurp — CHANGES_REQUESTED** [review]
Overall looks good, but there are some changes missing that I expected:

- [ ] `libgalois` was not changed to `libkatana`
- [ ] The main header is still `Galois.h`
- [ ] `README.md` and `Python.md` were not updated.
- [ ] Documentation prose does not appear to have been updated. Both doxygen and python doc strings.
- [ ] The python code in general seems to have some references to Galois like 
`GaloisCompiler`.

I'm fine with declaring some or all of this out of scope for this PR. But let's be explicit about it and create issues and/or todos to track the remaining things needed.

File: libgalois/include/katana/BufferedGraph.h
Comment (ddn0) [reply libgalois/include/katana/BufferedGraph.h:31]: Reordering includes. BufferedGraph relies on `boost::counting_iterator` but it did not include it, so it depended on includers of BufferedGraph.h including `counting_iterator.hpp` on its behalf; adding the dependency fixes this latent issue.

File: docs/introduction.dox
Comment (ddn0) [reply docs/introduction.dox:59]: My thinking is that there is the general project "katana" and libgalois, the (shared memory) parallel programming library. Katana also contains the analytics routines, partitioning library, etc.

Updating the docs should be done separately I think.

**Comment by ddn0** [comment]
> libgalois was not changed to libkatana

I was assuming that we'd leave libgalois as the name of the shared-memory parallel programming library.

> The main header is still Galois.h

Ditto

> README.md and Python.md were not updated.

Updating any human text is outside the scope of this change.

> Documentation prose does not appear to have been updated. Both doxygen and python doc strings.

Ditto.

> The python code in general seems to have some references to Galois like GaloisCompiler.

I kept things like `GaloisCompiler` and `numba_support.galois`  under the assumption that this exists to support the shared-memory library specifically. This was just my guess, so this is up for discussion certainly.

**Comment by arthurp** [comment]
> I was assuming that we'd leave libgalois as the name of the shared-memory parallel programming library.

I am quite happy with this distinction, in fact I recommended something like it a while ago, but I think we may have users who don't like it. @gurbinder533, you know more about the naming expectations of other people. Could you provide input?

> I kept things like `GaloisCompiler` and `numba_support.galois` under the assumption that this exists to support the shared-memory library specifically. This was just my guess, so this is up for discussion certainly.

You are right about `numba_support.galois`. `GaloisCompiler` would be used for distributed and even GPU algorithms written in Python. Though it isn't yet. Up to you if you want to change the name now to match the future usage.

**Review by arthurp — APPROVED** [review]
Once you have addressed the two remaining specific comments I have, LGTM. I recommend you talk to @gurbinder533 though about the `libgalois` issue. I'm adding him as a reviewer.

File: Python.md
Comment (gurbinder533) [inline Python.md:6]: "Galois for high-performance" --> "Katana for high-.."

File: Python.md
Comment (gurbinder533) [inline Python.md:8]: Galois --> Katana

File: Python.md
Comment (gurbinder533) [inline Python.md:43]: One more.

File: Python.md
Comment (gurbinder533) [inline Python.md:46]: One more.

File: README.md
Comment (gurbinder533) [inline README.md:171]: Galois/Lonestar --> katana/lonestar. 
The wording in this file needs to be changed a bit graph-convert does provide '-gr2kg' now but that can be done in a separate PR.

File: README.md
Comment (gurbinder533) [inline README.md:221]: libkatana?

File: README.md
Comment (gurbinder533) [inline README.md:224]: examples for Katana?

File: cmake/KatanaConfig.cmake.in
Comment (gurbinder533) [inline cmake/KatanaConfig.cmake.in]: Katana package?

File: docs/concurrent_data_structures.dox
Comment (gurbinder533) [inline docs/concurrent_data_structures.dox:18]: Galois also provides --> Katana also ...

File: docs/concurrent_data_structures.dox
Comment (gurbinder533) [inline docs/concurrent_data_structures.dox:79]: Katana provides two ....

File: docs/concurrent_data_structures.dox
Comment (gurbinder533) [inline docs/concurrent_data_structures.dox:84]: Katana label-computation graphs

**Comment by gurbinder533** [comment]
> Once you have addressed the two remaining specific comments I have, LGTM. I recommend you talk to @gurbinder533 though about the `libgalois` issue. I'm adding him as a reviewer.

I think we should go ahead and change `libgalois` to `libkatana` and `Galois.h` to `Katana.h` to avoid any type of confusion and not to have to explain what Galois means, especially since people writing applications in C++ will have to include "katana/Galois.h". Also, since namespaces are going to be `katana::`, it will just add to the confusion. This is what I think but we should discuss if I am missing anything. 

Also, please ignore my other comments on the PR on changing Galois to Katana in the text files. I should have read this discussion before leaving comments.

**Comment by ddn0** [comment]
> I think we should go ahead and change `libgalois` to `libkatana` and `Galois.h` to `Katana.h` to avoid any type of confusion and not to have to explain what Galois means, especially since people writing applications in C++ will have to include "katana/Galois.h". Also, since namespaces are going to be `katana::`, it will just add to the confusion. This is what I think but we should discuss if I am missing anything.

If we are going to subcomponents inside the katana project, I think having a `libkatana` might be confusing as well as "the katana library" may mean `libkatana` or all the relevant katana libraries.

If we follow the pattern of boost, we can have `#include katana/galois/Galois.h`, `#include katana/querying/....`, etc all within the `katana::` namespace.

I think the issue is that we have several related projects in the enterprise repo (querying, gnns, mining) that would benefit from shared code/build/packaging infrastructure but from the perspective of this repo, you don't see those projects, so subcomponents seems overly complicated for just `libtsuba`, `libgalois` and `libsupport`.

Thoughts?

**Comment by gurbinder533** [comment]
> I think the issue is that we have several related projects in the enterprise repo (querying, gnns, mining) that would benefit from shared code/build/packaging infrastructure but from the perspective of this repo, you don't see those projects, so subcomponents seems overly complicated for just `libtsuba`, `libgalois` and `libsupport`.
> 

Yeah, I agree that is precisely the issue. I was just thinking from the katana repo's perspective, it certainly makes more sense from the enterprise version. I am okay with keeping `libgalois` then.

**Comment by roshandathathri** [comment]
I strongly agree with Donald that we should NOT be changing `libgalois` to `libkatana`.

File: README.md
Comment (roshandathathri) [inline README.md:174]: What's the difference between katana and Katana?

File: cmake/KatanaConfig.cmake.in
Comment (roshandathathri) [reply cmake/KatanaConfig.cmake.in]: Agree

File: docs/thread_local_storage.dox
Comment (roshandathathri) [inline docs/thread_local_storage.dox]: Why are we picking `substrate` sub-namespace?

**Review by roshandathathri — COMMENTED** [review]
Have only seen 50 files so far. It would be good if you add a comment listing the rules you used to rename so that we can ensure we have a consensus on that.

**Comment by arthurp** [comment]
I think we have a lot of questions about what is called Galois and what is called Katana and even what is called katana (lower case). I would recommend we include documentation explaining the precise boundaries between Galois and Katana _in this PR_. For example, is the computational model the "Galois Model"?

File: README.md
Comment (tylershunt) [reply README.md:221]: I kind of like keeping the name for the core library `galois`; reserving `katana` as the the name for everything collectively.

**Review by tylershunt — APPROVED** [review]
Nice. With a shell script as well 😄 .  I think there are a lot of instances of the the `galois` brand in comments (libgalois has a license in every file). I'm not really sure how to handle those (or rather if we should handle them). I like keeping `galois` as a now more precise name for that library.

Some other stray comments with lingering galois but it seems like others have already addressed that. I would be fine merging  just to get the mechanical change in there.

**Comment by ddn0** [comment]
>  It would be good if you add a comment listing the rules you used to rename so that we can ensure we have a consensus on that.

There is a script that actually does the renaming `scripts/rename-galois-to-katana.sh` do you want something more than that?

**Comment by roshandathathri** [comment]
> There is a script that actually does the renaming `scripts/rename-galois-to-katana.sh` do you want something more than that?

No. That's great! Let me check that. All changes were made by the script?

**Comment by ddn0** [comment]
Yup, except for this one manual change (and running check_format -fix and check_ifndef -fix):

```diff
diff --git a/libgalois/include/katana/Allocators.h b/libgalois/include/katana/Allocators.h
index e6496f825..3c0fd5c0a 100644
--- a/libgalois/include/katana/Allocators.h
+++ b/libgalois/include/katana/Allocators.h
@@ -38,6 +38,8 @@
 #include "katana/substrate/PtrLock.h"
 #include "katana/substrate/SimpleLock.h"

+// TODO(ddn): Merge with Mem.h. Users should not include this file directly.
+
 namespace katana {

 extern unsigned activeThreads;
@@ -657,7 +659,8 @@ public:
 // Now adapt to standard std allocators
 ////////////////////////////////////////////////////////////////////////////////

-//! A fixed size block allocator
+//! Scalable fixed-sized allocator for T that conforms to STL allocator
+//! interface but does not support variable sized allocations
 template <typename Ty>
 class FixedSizeAllocator;

diff --git a/libgalois/include/katana/Mem.h b/libgalois/include/katana/Mem.h
index 60182d15e..1187d3611 100644
--- a/libgalois/include/katana/Mem.h
+++ b/libgalois/include/katana/Mem.h
@@ -48,11 +48,6 @@ typedef katana::BumpWithMallocHeap<katana::FreeListHeap<katana::SystemHeap>>
 typedef katana::ExternalHeapAllocator<char, IterAllocBaseTy> PerIterAllocTy;
 //! [PerIterAllocTy example]

-//! Scalable fixed-sized allocator for T that conforms to STL allocator
-//! interface but does not support variable sized allocations
-template <typename Ty>
-using FixedSizeAllocator = katana::FixedSizeAllocator<Ty>;
-
 //! Scalable variable-sized allocator for T that allocates blocks of sizes in
 //! powers of 2 Useful for small and medium sized allocations, e.g. small or
 //! medium vectors, strings, deques
```

I went back and forth about putting this in a separate commit, but I wanted to keep the state of the repo buildable modulo the actual renaming commit. Another option would be to put this patch into `scripts/rename-galois-to-katana.sh` if you think that would be helpful.

**Comment by roshandathathri** [comment]
> Another option would be to put this patch into `scripts/rename-galois-to-katana.sh` if you think that would be helpful.

No, this is fine. Thanks!

File: scripts/rename-galois-to-katana.sh
Comment (roshandathathri) [inline scripts/rename-galois-to-katana.sh]: Same comment as before. Why not do the same for `substrate`?

File: scripts/rename-galois-to-katana.sh
Comment (roshandathathri) [inline scripts/rename-galois-to-katana.sh]: FWIW I was fine with the previous directory structure

**Review by roshandathathri — APPROVED** [review]
LGTM

Better to merge asap.

**Review by gurbinder533 — APPROVED** [review]
LGTM.

**Comment by ddn0** [comment]
I've updated the PR description with the summary of the discussion so far, and I've resolved any issues that I think are still in scope. 

> I think we have a lot of questions about what is called Galois and what is called Katana and even what is called katana (lower case). I would recommend we include documentation explaining the precise boundaries between Galois and Katana in this PR. For example, is the computational model the "Galois Model"?

I'd like to keep this PR focused on the mechanical changes to files for machines. If we are happy with the general thrust implied here and summarized in the PR description, I can include some documentation as part of #17 . (FWIW, I don't think "Galois Model" should figure that prominently in the documentation versus "concepts to help understand using libgalois" or "concepts to help understand using libgluon".)

File: README.md
Comment (ddn0) [reply README.md:174]: I'd avoid this phrasing altogether and simply reference the graph convert tool itself. Anyway, all this conversion documentation and tooling is out of date because `.gr` files are deprecated/unsupported.

File: docs/thread_local_storage.dox
Comment (ddn0) [reply docs/thread_local_storage.dox]: I actually liked the separation between `galois::substrate` and `galois::`. The idea being user-code should never reference `galois::substrate` directly but it seems like a lot of code likes using `PerThreadStorage` so the idea of hiding `galois::substrate` seems untenable.

In the latest iteration of this PR, I've moved `galois::substrate` into `katana::`.

File: scripts/rename-galois-to-katana.sh
Comment (ddn0) [reply scripts/rename-galois-to-katana.sh]: flattened substrate too

File: docs/thread_local_storage.dox
Comment (roshandathathri) [reply docs/thread_local_storage.dox]: katana::internal serves this purpose right?

Comment (ddn0) [reply docs/thread_local_storage.dox]: Yeah, but my dream was always that PerThreadStorage would be `internal` :D

**Pull request merged**
